mlflow/tracing/export/mlflow.py

```diff
@@ -0,0 +1,146 @@
+                )
+            mlflow_span = span._to_mlflow_span()
+            self._trace_aggregator.add_span(mlflow_span)
+            mlflow_spans.append(mlflow_span)
+
+        # Call this after processing all spans because the parent-child order might
+        # not be preserved in the input spans
+        for span in mlflow_spans:
+            if span.parent_span_id is None:
+                self._export_trace(span)
+
+    def _export_trace(self, root_span: Span):
+        trace_data = self._trace_aggregator.pop_trace(root_span.context.trace_id)
+        if trace_data is None:
+            _logger.warning(f"Trace data with ID {root_span.context.trace_id} not found.")
+            return
+
+        # Create a TraceInfo object from the root span information
+        trace_info = TraceInfo(
+            trace_id=root_span.context.trace_id,
+            name=root_span.name,
+            start_time=root_span.start_time,
+            end_time=root_span.end_time,
+            status=root_span.status,
+            inputs=self._serialize_inputs_outputs(root_span.inputs),
+            outputs=self._serialize_inputs_outputs(root_span.outputs),
+            # TODO: These fields should be set if necessary
+            metadata={},
+            tags={},
+            source=None,
+        )
+        trace = Trace(trace_info, trace_data)
+        self._client.log_trace(trace)
+
+    def _serialize_inputs_outputs(self, input_or_output: Optional[Dict[str, Any]]) -> str:
+        """
+        Serialize inputs or outputs field of the span to a string, and truncate if necessary.
+        """
+        if not input_or_output:
+            return ""
+
+        try:
+            serialized = json.dumps(input_or_output)
+        except TypeError:
+            # If not JSON-serializable, use string representation
+            serialized = str(input_or_output)
+
+        return serialized[: self._MAX_CHARS_IN_TRACE_INFO_INPUTS_OUTPUTS]
+
+
+class InMemoryTraceDataAggregator:
+    """
+    Simple in-memory store for trace_id -> TraceData (i.e. spans).
+    """
+
+    _instance_lock = threading.Lock()
+    _instance = ContextVar("InMemoryTraceDataAggregator", default=None)
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance.get() is None:
+            with cls._instance_lock:
+                if cls._instance.get() is None:
+                    cls._instance.set(InMemoryTraceDataAggregator())
+        return cls._instance.get()
+
+    def __init__(self):
+        self._traces: Dict[str:TraceData] = {}
+        self._lock = threading.Lock()  # Lock for _traces
+
+    def add_span(self, span: Span):
+        if not isinstance(span, Span):
+            _logger.warning(f"Invalid span object {type(span)} is passed. Skipping.")
+            return
+
+        trace_id = span.context.trace_id
+        if trace_id not in self._traces:
+            with self._lock:
+                if trace_id not in self._traces:
```

Comment:
Why do we need another if check here?

Reply:
Trace might be added while a thread is waiting for lock at L129.

Let's say if two threads A and B trying to add spans with same trace id.
1. A reaches L128, evaluate False, and acquire lock.
2. B reaches L129, evaluate False, wait for lock.
3. A reaches L131, create a new trace data. Release the lock.
4. B acquire the lock. At this point, if we don't have the second check at L130, B will create a new trace data again.

Reply:
Yep reasonable!